src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/String/KeyAnalyzer.cs

```diff
@@ -263,25 +259,32 @@ public AnalysisResults(bool ignoreCase, bool allAsciiIfIgnoreCase, int hashIndex
             public bool RightJustifiedSubstring => HashIndex < 0;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ReadOnlySpan<char> SliceLeft(string s, int index, int count) => s!.AsSpan(index, count);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ReadOnlySpan<char> SliceRight(string s, int index, int count) => s!.AsSpan(s!.Length + index, count);
+
         private abstract class SubstringComparer : IEqualityComparer<string>
         {
             public int Index;
             public int Count;
-            public bool IsLeft;
+            public GetSpan Slicer = SliceLeft;
+
             public abstract bool Equals(string? x, string? y);
             public abstract int GetHashCode(string s);
         }
 
         private sealed class JustifiedSubstringComparer : SubstringComparer
         {
-            public override bool Equals(string? x, string? y) => x.AsSpan(IsLeft ? Index : (x!.Length + Index), Count).SequenceEqual(y.AsSpan(IsLeft ? Index : (y!.Length + Index), Count));
-            public override int GetHashCode(string s) => Hashing.GetHashCodeOrdinal(s.AsSpan(IsLeft ? Index : (s.Length + Index), Count));
+            public override bool Equals(string? x, string? y) => Slicer(x!, Index, Count).SequenceEqual(Slicer(y!, Index, Count));
```

Comment:
Here we can just call Slicer, which and no longer has an IsLeft test in a hot loop, so calling through to the `GetSpan`-like method that was set on lines 60 (for the `IsLeft == true` logic) and 84 (for the `IsLeft == false` logic). This (to my reasoning) will result in much better low-level branch-prediction and since the actual delegate bodies are ` [MethodImpl(MethodImplOptions.AggressiveInlining)]` they should be inlined stubs.

Reply:
I am *not* confident about the various not-null hints `!` that I injected, but I think they're right, if not overkill. Not sure it really would affect the code generated and welcome input on that from low-level ninjas